Create Configure Knative Networking page (#6518)
knative/docs: docs/versioned/serving/config-network-adapters.md

---
audience: administrator
components:
  - serving
function: how-to
---

# Configure Knative networking

This page provides installation and configuration guidance for Knative networking. You can configure Ingress controls, service-meshes, and gateways.

### Determine current state

Use the following command to determine which ingress controllers are installed and their status.

```bash
kubectl get pods -n knative-serving
```

The ingress controllers, that have been tested for Knative, have the following base names:

- Kourier: `kourier-control-*`, and `kourier-gateway-*`. Kourier is included in the Knative Serving installation should appear in the results when your cluster is first created.
- Contour: `contour-*`
- Istio: `istio-webhook-*`. The main Istio control plane pods such as `istiod-*` are in the `istio-system` namespace. Knative adds the `istio-webhook-*` pod in the `knative-serving` namespace when Istio is the chosen networking layer.

> The `network-config` ConfigMap sets which controller to use with the ingress controller key. This key is patched with the name of any new controller. See [Changing the ingress controller](#change-the-controller) for important information about using this key.

Reviewer: This is actually slightly different -- each ingress controller only manages ingress objects that are annotated with its key.  Knative Serving uses a default value of the key based on the `network-config` ConfigMap.  So this is somewhat of a cooperative process between the ingress controller and the core Knative controller.

Reviewer: (The end result is effectively the same, but the mechanism is different, which seemed important to correct in the documentation.)

## Network layer options

> Review the following tabs to determine the optimal networking layer for your cluster. For most users, the Kourier ingress controller is sufficient in conjunction the default Istio gateway, which is also included in the Knative Serving installation. You can expand your capabilities with the Contour ingress, a full-feature service mesh with Istio, and the Kubernetes Gateway API.

Reviewer: Folks shouldn't need _both_ Kourier and Istio -- one or the other should be sufficient, depending on need.

---
audience: administrator
components:
  - serving
function: how-to
---

# Configure Knative networking

> This page provides configuration guidance for Knative networking. You can configure Ingress controls, service-meshes, and gateways.

Reviewer: There are two parts of Kubernetes networking -- in cluster networking (which Knative doesn't need to do anything special with), and into-cluster networking (aka ingress), which Knative _does_ need to integrate with.  Clarify that this is for ingress networking:

```suggestion
This page provides configuration guidance for Knative Serving's integration with Kubernetes ingress controllers. Knative leverages existing ingress controls in your Kubernetes cluster, allowing you to use the same monitoring, .
```

- Certificate management: Configurable secrets for TLS encrypted traffic.
- Timeout policies: Controls for idle, and response stream timeouts.
- Traffic visibility: Mechanisms to expose services externally or cluster-locally.

The Knative `networking.internal.knative.dev` Ingress type is generally referred to as KIngress objects.

There are also third-party Knative networking options and Knative products available but are not tested or managed by the Knative community. For more information, see [Knative offerings](../install/knative-offerings.md).

=== "Kourier"

    Knative Serving network layer architecture:

    ```mermaid
    ---
    config:
      layout: elk
      theme: default
    ---
    flowchart LR
      subgraph top[" "]
        direction LR
            kingress1["Ingress object (KIngress)<br>networking.internal.knative.dev"]
            serving["Serving<br>controller"]
            route["Route object"]
      end
        route -- read by --> serving
        serving -- creates --> kingress1
        style kingress1 fill:#BBDEFB,stroke-width:1px,stroke-dasharray: 0
        style serving fill:#FFE0B2

>         style top fill:transparent

Reviewer: (Suggested where to add it above, and then remove from here, Contour, and Istio diagrams.  It's already missing from Gateway API.)

    ```

    Kourier network layer:

    ```mermaid
    ---
    config:
      layout: elk
      theme: default
    ---
    flowchart LR
     subgraph bottom[" "]
        direction LR
            envoy["Envoy deployment<br>kourier-system namespace"]
            kourier["net-kourier<br>controller"]
            kingress2["KIngress class:<br>kourier.ingress.networking.knative.dev"]
      end
        kingress2 -- read by --> kourier
        kourier -- programs --> envoy
    
        style envoy fill:#BBDEFB
        style kourier fill:#FFE0B2
        style bottom fill:transparent
    ```

    Kourier is a lightweight implementation of the KIngress resource and is suitable for clusters that don't need other ingress features. Kourier is optimal for learning and prototyping, and is installed by the Knative [Quickstart](../getting-started/README.md).

>     Kourier is is the only supported option for IBM-Z and IBM-P platforms. These IBM platforms require additional steps as documented in [Install Serving th YAML on IBM-Z and IBM-P](../install/yaml-install/serving/install-serving-with-yaml-on-IBM-Z-and-IBM-P.md).

Reviewer: ```suggestion
    Kourier is is the only supported option for IBM-Z and IBM-P platforms. These IBM platforms require additional steps as documented in [Install Serving with YAML on IBM-Z and IBM-P](../install/yaml-install/serving/install-serving-with-yaml-on-IBM-Z-and-IBM-P.md).
```

>     Kourier provides the following additional configuration options:

Reviewer: Do you want to link to documentation on any of these?

    ```mermaid
    ---
    config:
      layout: elk
      theme: default
    ---
    flowchart LR
      subgraph top[" "]
        direction LR
            kingress1["Ingress object (KIngress)<br>networking.internal.knative.dev"]
            serving["Serving<br>controller"]
            route["Route object"]
      end
        route -- read by --> serving
        serving -- creates --> kingress1
        style kingress1 fill:#BBDEFB,stroke-width:1px,stroke-dasharray: 0
        style serving fill:#FFE0B2
        style top fill:transparent
    ```

    Contour network layer:

    ```mermaid
    ---
    config:
      layout: elk
      theme: default
    ---
    flowchart LR
         subgraph bottom[" "]
            direction LR
            C1["KIngress objects"]
            C2("Knative<br>net-contour")
            C3["HTTPProxy<br>projectcontour.io"]
            C4("Contour")
          end
            C1 -- "read by" --> C2
            C2 -- "creates" --> C3
            C3 -- "read by" --> C4
        
            style C2 fill:#FFE0B2
            style C3 fill:#BBDEFB
            style C4 fill:#BBDEFB
            style bottom fill:transparent
    ```

>     The Contour ingress controller, `net-contour`, bridges Knative's KIngress resources to Contour's HTTPProxy resources. A good choice for clusters that already run non-Knative apps, are already using a Contour envoy, or don't need a full-feature service mesh.

Reviewer: ```suggestion
    The Contour ingress controller, `net-contour`, bridges Knative's KIngress resources to Contour's HTTPProxy resources. Contour is good choice for clusters that already run non-Knative apps, are already using a Contour envoy, or don't need a full-feature service mesh.
```

    Contour provides the following additional configuration options:

    - CORS policy configuration.

>     - Direct visibility classes for external and internal traffic.

Reviewer: Contour also supports a number of cluster-wide configuration options on its own: https://projectcontour.io/docs/1.33/configuration/

It's probably worth linking that here, so that it's clear that you can configure many options using the native Contour configuration.

=== "Istio"

    Knative Serving network layer architecture:

    ```mermaid
    ---
    config:
      layout: elk
      theme: default
    ---
    flowchart LR
      subgraph top[" "]
        direction LR
            kingress1["Ingress object (KIngress)<br>networking.internal.knative.dev"]
            serving["Serving<br>controller"]
            route["Route object"]
      end
        route -- read by --> serving
        serving -- creates --> kingress1
        style kingress1 fill:#BBDEFB,stroke-width:1px,stroke-dasharray: 0
        style serving fill:#FFE0B2
        style top fill:transparent
    ```

    Istio network layer:

    ```mermaid
    ---
    config:
      layout: elk
      theme: default
    ---
    flowchart LR
     subgraph bottom[" "]
        direction LR
            envoy["Envoy deployment<br>istio-system namespace"]
            kourier["net-istio<br>controller"]
            kingress2["KIngress class:<br>istio.ingress.networking.knative.dev"]
      end
        kingress2 -- read by --> kourier
        kourier -- programs --> envoy
    
        style envoy fill:#BBDEFB
        style kourier fill:#FFE0B2

>         style bottom fill:transparent

Reviewer: The pattern here is equivalent to Contour, except using the VirtualService resource type.  The diagram should be:

```suggestion
     subgraph bottom[" "]
        direction LR
            VirtualService["Istio<br>VirtualService"]
            istiod["Istio"]
            net-istio["net-istio<br>controller"]
            kingress["KIngress class:<br>istio.ingress.networking.knative.dev"]
      end
        kingress -- read by --> net-istio
        net-istio -- creates --> VirtualService
        VirtualService -- read by --> istiod
    
        style VirtualService fill:#BBDEFB
        style istiod fill:#BBDEFB
        style net-istio fill:#FFE0B2
        style bottom fill:transparent
```

    ```

>     The Knative `net-istio` is a KIngress controller for Istio. It's a full-feature service mesh that also functions as a Knative ingress. Well suited for enterprises already running Istio or who need advanced service mesh features.

Reviewer: ```suggestion
    The Knative `net-istio` is a KIngress controller for Istio. Istio is a full-feature service mesh that also functions as a Knative ingress. `net-istio` is well suited for enterprises already running Istio or who need advanced service mesh features.
```

    Knative has a default Istio integration without the full-feature service mesh. The `knative-ingress-gateway` in the `knative-serving` namespace is a shared Istio gateway resource that handles all incoming (north-south) traffic to Knative services. This gateway points to the underlying `istio-ingressgateway` service in the `istio-system` namespace. You can replace this gateway with one of your own. See [Configuring the Ingress gateway](setting-up-custom-ingress-gateway.md).

    Istio provides the following additional configuration options:

    - Advanced gateway selection with label selectors for fine-grained routing.

>     - Support for mesh-aware and cluster-local access.

Reviewer: ```suggestion
    In addition to [Istio's configuration options](https://istio.io/latest/docs/reference/config/), `net-istio` provides the following additional configuration options for mapping Knative Routes to Istio VirtualServices:

    - Advanced gateway selection with label selectors for fine-grained routing.
    - Support for mesh-aware and cluster-local access.
```

=== "Gateway API"

    ```mermaid
    ---
    config:
      layout: elk
      theme: default
    ---
    flowchart LR
     subgraph net-gateway-api["net-gateway-api&nbsp;controller"]
            GW["Gateway"]
            Route["Knative&nbsp;Route"]
            HR["HTTPRoute"]
      end
     subgraph underlying["Underlying&nbsp;Controller<br>(Contour │ Istio │ Envoy Gateway │ …)"]
        Controller["GatewayClass&nbsp;Controller"]
      end
        KSvc["Knative&nbsp;Service"] --> Route
        Route -- translates&nbsp;to --> GW & HR
        GW --> Controller
        HR --> Controller
        Controller -- routes&nbsp;traffic&nbsp;to --> Pods["Your&nbsp;Pods"]

    style net-gateway-api fill:#e3f2fd,stroke:#1976d2
    style underlying fill:#fff3e0,stroke:#ef6c00
    ```

>     The Knative `net-gateway-api` is a KIngress implementation for Knative integration with the [Kubernetes Gateway API](https://gateway-api.sigs.k8s.io/). A good choice for teams adopting the Gateway API to unify ingress across Kubernetes.

Reviewer: I would say that Gateway API is recommended for many teams, but may experience limits with large numbers of Knative Services.  @dprotaso can correct me here.

```yml
ingress-class: contour.ingress.networking.knative.dev
ingress.class: kourier.ingress.networking.knative.dev
```

If you want to switch to a controller that is already installed, patch the `config-network` ConfigMap with the new controller as shown in the following example using Kourier.

```bash
kubectl patch cm config-network -n knative-serving \
  --type merge -p '{"data":{"ingress-class":"kourier.ingress.networking.knative.dev"}}'
```

You can remove an unused key with a dot with the following command:

```bash
kubectl patch configmap config-network -n knative-serving \                                                    
--type=json -p='[{"op": "remove", "path": "/data/ingress.class"}]'
```

If you want to install a new controller, see [Install serving with YAML](../install/yaml-install/serving/install-serving-with-yaml.md). Be aware that changing the Ingress class of an existing Route may result in undefined behavior.

## Gateway configurations

Knative assumes two gateways in the cluster:

- Externally exposed - defines the gateway for external traffic.

> - Internally-only exposed - defines the gateway for local traffic.

Reviewer: Can you add a TODO here to add a diagram?  I don't think we need to add it now, but it would be a handy future addition.

    Knative has a default Istio integration without the full-feature service mesh. The `knative-ingress-gateway` in the `knative-serving` namespace is a shared Istio gateway resource that handles all incoming (north-south) traffic to Knative services. This gateway points to the underlying `istio-ingressgateway` service in the `istio-system` namespace. You can replace this gateway with one of your own. See [Configure Istio's ingress gateway](setting-up-custom-ingress-gateway.md).

    Istio provides the following additional configuration options:

    - Advanced gateway selection with label selectors for fine-grained routing.
    - Support for mesh-aware and cluster-local access.
    
    In addition to [Istio's configuration options](https://istio.io/latest/docs/reference/config/), `net-istio` provides the following additional configuration options for mapping Knative Routes to Istio VirtualServices:

    - Advanced gateway selection with label selectors for fine-grained routing.
    - Support for mesh-aware and cluster-local access.

=== "Gateway API"

    ```mermaid
    ---
    config:
      layout: elk
      theme: default
    ---
    flowchart LR

>      net-gateway["net-gateway-api controller"]

Reviewer: ```suggestion
     net-gateway("net-gateway-api<br>controller")
```

Let's narrow this text so we can get slightly bigger font here.  Also, let's use rounded-boxes for software components like controllers, and square boxes for Kubernetes resources (objects).

     GW["Gateway"]
     KIngress["KIngress objects"]
     HR["HTTPRoute"]
     subgraph underlying["Gateway-API Implementation<br>(Contour, Istio, Envoy Gateway, …)"]
        Controller("GatewayClass&nbsp;Controller")
      end
        KIngress -- read by --> net-gateway
        net-gateway -- updates --> GW
        net-gateway -- creates --> HR
        GW -- read by --> Controller
        HR -- read by --> Controller
    style net-gateway fill:#FFE0B2
    style GW fill:#BBDEFB
    style HR fill:#BBDEFB
    style Controller fill:#BBDEFB
    style underlying fill:#fff3e0,stroke:#ef6c00
    ```

    The Knative `net-gateway-api` is a KIngress implementation for Knative integration with the [Kubernetes Gateway API](https://gateway-api.sigs.k8s.io/). A recommended choice for many teams adopting the Gateway API to unify ingress across Kubernetes. Clusters with large numbers of Knative Services however, may experience limitations.

>     The Kubernetes Gateway API requires a controller or service mesh. Istio and Contour implementations are tested. For more information see [Tested Gateway API version and Ingress](https://github.com/knative-extensions/net-gateway-api/blob/main/docs/test-version.md).

Reviewer: Service meshes and controllers are _different_ types of things.  It would be like saying "gardens need fruit or paths".  Let's call them implementations, and include envoy-gateway.

```bash
kubectl patch cm config-network -n knative-serving \
  --type merge -p '{"data":{"ingress-class":"kourier.ingress.networking.knative.dev"}}'
```

You can remove an unused key with a dot with the following command:

```bash
kubectl patch configmap config-network -n knative-serving \                                                    
--type=json -p='[{"op": "remove", "path": "/data/ingress.class"}]'
```

If you want to install a new controller, see [Install serving with YAML](../install/yaml-install/serving/install-serving-with-yaml.md). Be aware that changing the Ingress class of an existing Route may result in undefined behavior.

In scenarios where there are multiple networking implementations, you can create a custom ingress class to specify different ingress class annotations for each service. For more information, see [Configuring Services custom ingress class](./services/ingress-class.md).

There are also third-party Knative networking options and Knative products available but are not tested or managed by the Knative community. For more information, see [Knative offerings](../install/knative-offerings.md).

> ## Gateway configurations

Reviewer: This is all Gateway-API specific.  I don't know if you want to create a parallel page with "Configure Istio's ingress gateway" (e.g. "Configure Gateway API's ingress gateway"), or if you just want to add a clarification banner here that the following section only applies to Gateway API.

Author: Your call, I can create a new page - would it need additional content?

The Knative tested ingress controllers (Contour, Istio, and Kourier) have the following common configurations:

- Certificate management: Configurable secrets for TLS encrypted traffic. See [Using a custom TLS certificate for DomainMapping](./services/custom-tls-certificate-domain-mapping.md).
- Timeout policies: Controls for idle, and response stream timeouts. See [Configuring the Defaults ConfigMap](./configuration/config-defaults.md) to review timeout settings.
- Traffic visibility: Mechanisms to expose services externally or cluster-locally. See [Traffic management](./traffic-management.md).
- The mapping of Route objects to `networking.internal.knative.dev` objects.

The Knative `networking.internal.knative.dev` Ingress type is generally referred to as KIngress objects.

> In scenarios where there are multiple networking implementations, you can create a custom ingress class to specify different ingress class annotations for each service. For more information, see [Configuring Services custom ingress class](./services/ingress-class.md)

Author: FYI - I added this link to the custom ingress class (new).

## Network layer options

Review the tabbed content in this section to determine the optimal networking layer for your cluster. If you already have one of the ingress controllers installed in your cluster, we recommend using your existing installation. For most users without a supported ingress, the Kourier ingress controller is sufficient. You can expand your capabilities with the Contour ingress, a full-feature service mesh with Istio, and the Kubernetes Gateway API.

The Knative tested ingress controllers (Contour, Istio, and Kourier) have the following common configurations:

- Certificate management: Configurable secrets for TLS encrypted traffic. See [Using a custom TLS certificate for DomainMapping](./services/custom-tls-certificate-domain-mapping.md).
- Timeout policies: Controls for idle, and response stream timeouts. To review timeout settings, see [Configuring the Defaults ConfigMap](./configuration/config-defaults.md).
- Traffic visibility: Mechanisms to expose services externally or cluster-locally. See [Traffic management](./traffic-management.md).
- The mapping of Route objects to `networking.internal.knative.dev` objects is common to these controllers.

> The Knative `networking.internal.knative.dev` Ingress type is generally referred to as KIngress objects.

Author: A great developmental edit observation.

    The Kubernetes Gateway API requires a controller or service mesh. Istio and Contour implementations are tested. For more information see [Tested Gateway API version and Ingress](https://github.com/knative-extensions/net-gateway-api/blob/main/docs/test-version.md).

## Ingress configurations

Use the following command to determine which ingress controllers are installed and their status.

``` bash
kubectl get pods -n knative-serving
```

The ingress controllers have the following names:

- Kourier: `kourier-control-*`, and `kourier-gateway-*`. Kourier is included in the Knative Serving installation should appear in the results when your cluster is first created.
- Contour: `contour-*`
- Istio: `istio-webhook-*`. The main Istio control plane pods such as `istiod-*` are in the `istio-system` namespace. Knative adds the `istio-webhook-*` pod in the `knative-serving` namespace when Istio is the chosen networking layer.

> 

Reviewer: ```suggestion
- Gateway API: `net-gateway-api-controller` is the deployment used to interface with the Gateway API. It uses the configuration from `config-gateway` to select existing Gateway resources on the cluster.

```

kubectl patch cm config-network -n knative-serving \
  --type merge -p '{"data":{"ingress-class":"kourier.ingress.networking.knative.dev"}}'
```

You can remove an unused key with a dot with the following command:

```bash
kubectl patch configmap config-network -n knative-serving \                                                    
--type=json -p='[{"op": "remove", "path": "/data/ingress.class"}]'
```

If you want to install a new controller, see [Install serving with YAML](../install/yaml-install/serving/install-serving-with-yaml.md). Be aware that changing the Ingress class of an existing Route may result in undefined behavior.

## Gateway configurations

Knative assumes two gateways in the cluster:

- Externally exposed - defines the gateway for external traffic.
- Internally exposed - defines the gateway for local traffic.

Both gateways route traffic to the Knative services inside the cluster.

```mermaid
---
config:
  look: neo
  theme: redux
  layout: elk
---
flowchart LR
 subgraph Cluster["Knative Cluster"]
    direction TB
        ExtGateway["External Gateway<br>Exposed to external traffic"]
        IntGateway["Internal Gateway<br>Only for local internal traffic"]

>         KServe["Knative Services<br>Serving and Eventing"]

Reviewer: ```suggestion
        KServe["Knative Services"]
```

This is all Serving.  Unfortunately, Eventing has a somewhat different model and doesn't have clear abstractions for exposing the event ingress from outside the cluster.

  end
    ExtGateway -- Routes external HTTP/HTTPS traffic --> KServe
    IntGateway -- Routes internal traffic --> KServe
    ExternalClient["External Client<br>Internet / Outside cluster"] -- Ingress --> ExtGateway

>     InternalClient["Internal Client<br>Pod / Service inside cluster"] -- "Cluster-internal" --> IntGateway

Reviewer: "InternalClient" needs to be in the "Knative Cluster" subgraph.

    style ExtGateway fill:#ffe4e1,stroke:#ff0000,stroke-width:2px
    style IntGateway fill:#e6ffe6,stroke:#008000,stroke-width:2px
    style ExternalClient fill:#f0f8ff,stroke:#333
    style InternalClient fill:#f0fff0,stroke:#333
    style Cluster stroke:#333,stroke-width:3px,stroke-dasharray: 5 5
```    

Use the following command to list Kubernetes GatewayClass resources cluster-wide.

```bash
kubectl get gatewayclass
```

Example results:

```bash
NAME           CONTROLLER                    ACCEPTED   AGE
istio          istio.io/gateway-controller   True       11d
istio-remote   istio.io/unmanaged-gateway    True       11d
```

> When gateways are installed, the `config-gateway` ConfigMap is updated to track the `class`, `gateway`, and `service`. The service is the Kubernetes Service name that points to the pods in the Gateway implementation.

Author: The "Edit the `config-gateway` ConfigMap" step in netadapter-gateway.api (in the installation instructions) show editing the YAML for the external and internal clusters. Does that step suffice? Should it be repeated here?
